src/libraries/System.Private.CoreLib/src/System/Number.DecimalIeee754.cs

```diff
@@ -0,0 +1,255 @@
+            {
+                Signed = signed;
+                Exponent = exponent;
+                Significand = significand;
+            }
+        }
+
+        internal static DecimalIeee754<TSignificand> UnpackDecimalIeee754<TDecimal, TSignificand, TValue>(TValue value)
+            where TDecimal : unmanaged, IDecimalIeee754UnpackInfo<TDecimal, TSignificand, TValue>
+            where TSignificand : IBinaryInteger<TSignificand>
+            where TValue : IBinaryInteger<TValue>
+        {
+            bool signed = (value & TDecimal.SignMask) != TValue.Zero;
+            TSignificand significand;
+            int exponent;
+
+            if ((value & TDecimal.G0G1Mask) == TDecimal.G0G1Mask)
+            {
+                exponent = TDecimal.ConvertToExponent((value & TDecimal.G2ToGwPlus3ExponentMask) >> (TDecimal.NumberBitsSignificand + 1));
+                significand = TDecimal.ConvertToSignificand((value & TDecimal.GwPlus4SignificandMask) | TDecimal.MostSignificantBitOfSignificandMask);
+            }
+            else
+            {
+                exponent = TDecimal.ConvertToExponent((value & TDecimal.G0ToGwPlus1ExponentMask) >> (TDecimal.NumberBitsSignificand + 3));
+                significand = TDecimal.ConvertToSignificand(value & TDecimal.GwPlus2ToGwPlus4SignificandMask);
+            }
+
+            return new DecimalIeee754<TSignificand>(signed, exponent - TDecimal.Bias, significand);
+        }
+
+        internal static int CompareDecimalIeee754<TDecimal, TSignificand, TValue>(TValue currentValue, TValue otherValue)
+            where TDecimal : unmanaged, IDecimalIeee754UnpackInfo<TDecimal, TSignificand, TValue>
+            where TSignificand : IBinaryInteger<TSignificand>
+            where TValue : IBinaryInteger<TValue>
+        {
+            if (currentValue == otherValue)
+            {
+                return 0;
+            }
+            DecimalIeee754<TSignificand> current = UnpackDecimalIeee754<TDecimal, TSignificand, TValue>(currentValue);
+            DecimalIeee754<TSignificand> other = UnpackDecimalIeee754<TDecimal, TSignificand, TValue>(otherValue);
+
+            if (current.Signed && !other.Signed) return -1;
+
+            if (!current.Signed && other.Signed) return 1;
+
+            if (current.Exponent > other.Exponent)
+            {
+                return current.Signed ? -InternalUnsignedCompare(current, other) : InternalUnsignedCompare(current, other);
+            }
+
+            if (current.Exponent < other.Exponent)
+            {
+                return current.Signed ? InternalUnsignedCompare(other, current) : -InternalUnsignedCompare(current, other);
+            }
+
+            if (current.Significand == other.Significand) return 0;
+
+            if (current.Significand > other.Significand)
+            {
+                return current.Signed ? -1 : 1;
+            }
+            else
+            {
+                return current.Signed ? 1 : -1;
+            }
```

Comment:
This looks like it could be simplified similarly to doing a swap and `TSignificand.Compare(current, other)`